### PR TITLE
Add pause and resume

### LINKED ISSUE #85: Pausing in the console

### PR DESCRIPTION
**Pull request by seanmonstar**
A first step to address #70. You can now press the spacebar to "pause" the console. The subscriber will continue to process events, but as along as the client is "paused", it won't send any task updates. Resuming will bring the client back to "live", getting all updates again.

paused

I've tried a few different UI things, from making the whole background of the top red, to just making the word red... It's fine if others feel there's a better or prettier way to show this, but to keep moving forward, I just picked something that seemed good enough.

Closes #85

### CODE REVIEW

File: console-subscriber/src/aggregator.rs [hunk 1]
```diff
@@ -67,6 +70,12 @@ pub(crate) struct Flush {
     pub(crate) triggered: AtomicBool,
 }
 
+#[derive(Debug)]
+enum Temporality {
```
Comment (hawkw) [inline console-subscriber/src/aggregator.rs]: cute name :)

assuming this is an enum because we might want to add replays or other forms of time travel?

File: console-subscriber/src/aggregator.rs [hunk 2]
```diff
@@ -59,6 +59,9 @@ pub(crate) struct Aggregator {
 
     /// Map of task IDs to task stats.
     stats: TaskData<Stats>,
+
+    /// The time "state" of the aggregator, such as paused or live.
+    temporality: Temporality,
```
Comment (hawkw) [inline console-subscriber/src/aggregator.rs]: Having one temporality for the aggregator seems like a potential issue. Since multiple consoles can connect to the same application, if one user pauses their console, sending the pause command to the application, all consoles will be paused. Similarly, if I unpause my console, everyone else's console unpauses, too. It doesn't _seem_ like this is the behavior we want when multiple consoles are connected?

If we're going to implement pausing in the aggregator-side (rather than in the UI-side), it seems like we would want to track a separate temporality for each client connection, and decide whether we are going to send updates to each client independently?

File: proto/tasks.proto [hunk 3]
```diff
@@ -9,6 +9,8 @@ import "common.proto";
 service Tasks {
     rpc WatchTasks(TasksRequest) returns (stream TaskUpdate) {}
     rpc WatchTaskDetails(DetailsRequest) returns (stream TaskDetails) {}
+    rpc Pause(PauseRequest) returns (PauseResponse) {}
+    rpc Resume(ResumeRequest) returns (ResumeResponse) {}
 }
```
Comment (hawkw) [inline proto/tasks.proto:14]: Hmm, as per my comment above, I think if we want to implement pausing in the aggregator rather than in the UI, we would want to have a separate temporality for each client connection, rather than adding new RPCs that globally mutate whether the app is paused or resumed.

IMO, the best way to do this would be to turn the `WatchTasks` and `WatchTaskDetails` RPCs into bidirectional streaming RPCs, like
```suggestion
service Tasks {
    rpc WatchTasks(stream WatchCommand) returns (stream TaskUpdate) {}
    rpc WatchTaskDetails(stream WatchCommand) returns (stream TaskDetails) {}
}

message WatchCommand {
    enum Kind {
        PAUSE = 0;
        RESUME = 1;
    } 
    Kind command_kind = 1;
}
```
or something? Then, when a `PAUSE` message is sent on a given RPC, the aggregator would pause sending updates to that client until it receives a `RESUME` command from the same client.

**Review by hawkw — COMMENTED** [review]
Would you mind explaining the rationale for implementing this in the subscriber, rather than purely in the console? 

After reading the PR description, my naive assumption was that pausing would work by having the console UI maintain two copies of the state, one that's being updated as new data comes in, and one that's being displayed. When the console is paused, in this design, we would freeze the current state, duplicate it, and continue updating the duplicate rather than the displayed state; when unpausing, we would replace the displayed state with the updated state. Without having thought about it a great deal, this _seems_ like a simpler approach with fewer moving parts (and a conceptually cleaner separation of concerns), so I'd love to know if there's something I'm overlooking.

**Comment by seanmonstar** [comment]
I did spend a bit of time thinking about the different ways to implement this:

- In the console app, but the downsides are that you can't fetch "details" (like task or resources) in the paused state, since the subscriber has kept going. It also means that new viewing applications, like perhaps a web view, need to re-implement time travel entirely.
- In the subscriber app, but each pause received by a client means the state gets cloned into a frozen version. It's a bit more complicated to support, and also more "expensive".
- As this PR does, the entire subscriber state is affected by any client commands. This means multiple clients can't have separate state.

After discussing with others and some thinking, I believe the ability to support multiple clients is just because of the simple design of the current server. It doesn't seem like an actual requirement or use case that people would need. Customer interviews didn't show they expected to have multiple members of their team all connecting to the same instance. Anyone who did want to explore a process individually could copy an eventual recording file and play it over locally.

Does it seem that would be something users would likely expect to do? We certainly *can* support it, but it's more complicated in code and and costs more in CPU and memory.

**Comment by hawkw** [comment]
> * In the console app, but the downsides are that you can't fetch "details" (like task or resources) in the paused state, since the subscriber has kept going. It also means that new viewing applications, like perhaps a web view, need to re-implement time travel entirely.
> 
> * In the subscriber app, but each pause received by a client means the state gets cloned into a frozen version. It's a bit more complicated to support, and also more "expensive".

Hmm, I still think that cloning the state in the client is the "correct" thing, since you can continue to fetch details. My thought process is that it's better to make the console app do as much of the "expensive" work as possible, to support debugging in prod — there is, of course, going to be some performance hit on the instrumented application no matter what we do, but it would be nice to minimize it as much as possible, and make the console app do the expensive work.

I'm also not entirely convinced that "new viewing applications, like perhaps a web view, need to re-implement time travel entirely" is necessarily all that bad a downside, since there's already a lot of stuff that a web view or other UI would have to re-implement.

I guess in general, my bias is for putting as much intelligence in the UI as possible, and having the instrumented application just focus on doing anything that _has_ to happen in the app (e.g. anchoring timestamps) and getting the data out as efficiently as possible. But, I'm okay with going against that bias when it makes sense, situationally.

> After discussing with others and some thinking, I believe the ability to support multiple clients is just because of the simple design of the current server. It doesn't seem like an actual requirement or use case that people would need. Customer interviews didn't show they expected to have multiple members of their team all connecting to the same instance. Anyone who did want to explore a process individually could copy an eventual recording file and play it over locally.
> 
> Does it seem that would be something users would likely expect to do? We certainly _can_ support it, but it's more complicated in code and and costs more in CPU and memory.

Hmm...to me, it seemed like a reasonable use case we might want to be able to support, but if you've actually talked to people and nobody really wants it, I'm willing to be convinced otherwise. I will say, though, that the future uses of multiple clients I was imagining was less about two or more human users interactively debugging the same process, and more about a human user interactively debugging a process while data from that process is also being collected/aggregated/archived by some automated tool. but, that doesn't exist yet, so maybe it's a problem we only need to solve when those tools actually exist...

FWIW I'd be interested to hear more about the conversations you've had with potential users, if you don't mind sharing?

----

In summary: I'm fine to move forward with this approach if you're convinced it's the best thing, I just wanted to make sure we talked through it first.

**Comment by seanmonstar** [comment]
>  if you've actually talked to people and nobody really wants it, I'm willing to be convinced otherwise. [..] I'd be interested to hear more about the conversations you've had with potential users, if you don't mind sharing?

We showed something very similar to [this demo](https://www.youtube.com/watch?v=JGCewPUvF70) live to a few teams using Rust, and then asked what more they'd want to see, and how they'd expect to use it. [This document](https://hackmd.io/6xPTpWK5RGO_eGygI9QEPA?view) describes the feature work that we understood they wanted. (I realized this doesn't include the original conversations...)

> the future uses of multiple clients I was imagining was less about two or more human users interactively debugging the same process, and more about a human user interactively debugging a process while data from that process is also being collected/aggregated/archived by some automated tool.

The second part of the document I just linked mentions recording the events automatically (well, with a config option), and that would continue to happen "live".

> my bias is for putting as much intelligence in the UI as possible, and having the instrumented application just focus on doing anything that has to happen in the app (e.g. anchoring timestamps) and getting the data out as efficiently as possible.

This was my initial feeling too... And then it seemed somewhere along the way, the subscriber started to hold most of the state, and the console just "viewed" it. I've tried to go "remember" why that is, and it looks like it was trying to cut down on the amount of IO from the subscriber. But in order to support more of these temporal features, such as recording and stepping backward (rewind), we'll _need_ to save the individual events somewhere anyways...

I now am wondering about that decision. Perhaps sending a write/packet per event is the bad part, but just buffering a reasonable amount, and/or maybe  enabling gRPC compression (tonic now has it), would make the cost small enough that we don't need to really keep any state in the subscriber...

**Comment by hawkw** [comment]
> I now am wondering about that decision. Perhaps sending a write/packet per event is the bad part, but just buffering a reasonable amount, and/or maybe enabling gRPC compression (tonic now has it), would make the cost small enough that we don't need to really keep any state in the subscriber...

Probably we should do some actual testing to try and figure out how much data actually goes on the wire in a real program; it would be nice to have some data points for what is basically reasonable.

**Comment by hawkw** [comment]
> > if you've actually talked to people and nobody really wants it, I'm willing to be convinced otherwise. [..] I'd be interested to hear more about the conversations you've had with potential users, if you don't mind sharing?
> 
> We showed something very similar to [this demo](https://www.youtube.com/watch?v=JGCewPUvF70) live to a few teams using Rust, and then asked what more they'd want to see, and how they'd expect to use it. [This document](https://hackmd.io/6xPTpWK5RGO_eGygI9QEPA?view) describes the feature work that we understood they wanted. (I realized this doesn't include the original conversations...)

This doc is really nice, btw — I realize now you had posted a link in #70, but I think I'd overlooked that at the time. It would be nice to get some of the various features proposed there written up as separate issues, eventually.

**Comment by hawkw** [comment]
@seanmonstar any updates on this?

**Comment by seanmonstar** [comment]
So, I think it'd be good to carve out an issue to think about state in the subscriber vs the console. I suspect we can have the subscriber spit out events similar to the way it records to a file, and have the console build up the state, as long as we do the right kind of buffering.

But I also think this approach for pausing works at the moment, and the "cost" only exists if you actually connect and pause. We'd probably need to move pausing entirely to the console, but I think to do so would need answering the bigger question in the previous paragraph. So, if that seems fine, I can fixup the conflicts and open a separate issue around general state. Whatcha think?

**Review by hawkw — APPROVED** [review]
I'm fine to roll forwards with merging this once the merge conflicts are fixed :)
